File: src/api/role.ts
export * from './role/expert'
import { invoke } from '@tauri-apps/api/tauri'
import { invokeWithFriendlyError } from './helpers'
import type {
  PluginBackends,
  PluginBackendsOverride,
  PluginBackendsSourceMap,
} from './settings'

export interface UserRelationDto {
  id: string
  name: string
  prompt_hint: string
  favor_multiplier: number
  /** Role pack configured initial favorability (0–100); switching top-bar identity re-seeds favorability */
  initial_favorability: number
}

/** Virtual-time schedule inference from manifest `life_schedule` */

export interface LifeStateDto {
  label: string
  activity_key: string
  busy_level: number
  preferred_scene_id: string | null
}

/** Matches `plugin_backends.directory_plugins` snake_case JSON fields */

export type PersonalitySource = 'vector' | 'profile'

/** Normalized role pack root `ui.json`; aligned with frontend `UiConfig`; slot keys use dots */

export interface PackUiSlotConfig {
  order: string[]
  visible: string[]
}


export interface PackUiSlots {
  'chat_toolbar': PackUiSlotConfig
  'settings.panel': PackUiSlotConfig
  'role.detail': PackUiSlotConfig
  'sidebar': PackUiSlotConfig
  'chat.header': PackUiSlotConfig
}


export interface PackUiTheme {
  primaryColor?: string
  backgroundColor?: string
  fontFamily?: string
}


export interface PackUiLayout {
  sidebar?: string
  chatInput?: string
}


export interface PackUiConfig {
  shell: string
  theme: PackUiTheme
  layout: PackUiLayout
  slots: PackUiSlots
}


export interface AuthorRecommendedPlugin {
  id: string
  version_range?: string | null
  slots?: string[]
  for_backends?: string[]
  optional?: boolean
  note?: string | null
}


export interface AuthorPackFile {
  schema_version: number
  summary?: string
  detail_markdown?: string
  recommended_plugins?: AuthorRecommendedPlugin[]
  suggested_ui?: PackUiConfig | null
  suggested_plugin_backends?: PluginBackends | null
}


export interface RoleData {
  role_id: string
  name: string
  version: string
  author: string
  description: string
  personality_vector: number[]
  current_favorability: number
  current_emotion: string
  memory_count: number
  event_count: number
  user_relations: UserRelationDto[]
  default_relation: string
  relation_state: string
  current_user_relation: string
  /** Whether "default identity" is selected (manifest `default_relation`) */
  use_manifest_default: boolean
  /** Remote life enabled (DB) */
  remote_life_enabled: boolean
  /** Role pack suggested default for remote life (`settings.json` → `remote_presence.default_enabled`) */
  remote_life_pack_default: boolean | null
  event_impact_factor: number
  /** `evolution.personality_source`; defaults to vector */
  personality_source?: PersonalitySource
  /** Effective model: manifest `OLLAMA_MODEL` or user default */
  effective_ollama_model: string
  /** Global identity vs per-scene (`manifest identity_binding`) */
  identity_binding: 'global' | 'per_scene'
  /** Interaction mode in DB: `immersive` | `pure_chat` */
  interaction_mode: 'immersive' | 'pure_chat'
  /** Suggested default from role pack settings; may be null */
  interaction_mode_pack_default: 'immersive' | 'pure_chat' | null
  /** Virtual-time schedule inference from life placement; null when none */
  current_life: LifeStateDto | null
  /** Module subsystem backends resolved by `PluginHost` */
  plugin_backends: PluginBackends
  /** Session conversation override; null when none */
  plugin_backends_session_override?: PluginBackendsOverride | null
  /** Effective backends after session override */
  plugin_backends_effective?: PluginBackends
  /** Effective backend sources (pack/session/env) */
  plugin_backends_effective_sources?: PluginBackendsSourceMap
  /** Role pack `ui.json` shell / theme / layout / slots */
  pack_ui_config: PackUiConfig
  /** `author.suggested_ui` or role pack baseline UI; frontend `pack_ui_baseline` */
  pack_ui_baseline?: PackUiConfig
  /** Optional `author.json` */
  author_pack?: AuthorPackFile | null
  /** v2 blueprint `slot_registry`; null for legacy */
  slot_registry_pack?: import('../lib/slotRegistry').SlotRegistryMap | null
  slot_registry_effective?: import('../lib/slotRegistry').SlotRegistryMap | null
  slot_session_overridden_keys?: string[]
  blueprint_groups_pack?: import('../lib/slotRegistry').SlotGroupsMap | null
}


export interface SceneLabelEntry {
  id: string
  label: string
}

/** Flat UI snapshot from `get_role_info` / `switch_scene`; identity fields match {@link RoleData} */

export interface RoleInfo {
  role_id: string
  role_name: string
  version: string
  author: string
  description: string
  current_favorability: number
  current_emotion: string
  personality_vector: number[]
  /** `evolution.personality_source`; defaults to vector */
  personality_source?: PersonalitySource
  last_interaction?: string | null
  scenes: string[]
  /** Labels in `scenes` order; from role pack `scene.json` or manifest */
  scene_labels: SceneLabelEntry[]
  current_scene: string | null
  /** User narrative scene in DB; may differ from `current_scene` */
  user_presence_scene: string | null
  virtual_time_ms: number
  user_relations: UserRelationDto[]
  default_relation: string
  current_user_relation: string
  /** Whether "default identity" is selected; sentinel is `OCLIVE_DEFAULT_RELATION_SENTINEL` */
  use_manifest_default: boolean
  /** Relation stage from `role_runtime.relation_state` */
  relation_state: string
  remote_life_enabled: boolean
  remote_life_pack_default: boolean | null
  event_impact_factor: number
  /** Effective model: manifest `OLLAMA_MODEL` or user default */
  effective_ollama_model: string
  /** Global identity vs per-scene (`manifest identity_binding`) */
  identity_binding: 'global' | 'per_scene'
  interaction_mode: 'immersive' | 'pure_chat'
  interaction_mode_pack_default: 'immersive' | 'pure_chat' | null
  /** Virtual-time schedule inference; null when none */
  current_life: LifeStateDto | null
  /** Module subsystem backends from `load_role` snapshot */
  plugin_backends: PluginBackends
  /** Session override; null when none */
  plugin_backends_session_override?: PluginBackendsOverride | null
  /** Effective backends after session override */
  plugin_backends_effective?: PluginBackends
  /** Effective backend sources (pack/session/env) */
  plugin_backends_effective_sources?: PluginBackendsSourceMap
  /** Whether worldview knowledge index is loaded from disk */
  knowledge_enabled?: boolean
  /** Knowledge chunk count in loaded index; 0 when disabled */
  knowledge_chunk_count?: number
  /** Role pack `ui.json` shell / theme / layout / slots */
  pack_ui_config: PackUiConfig
  pack_ui_baseline?: PackUiConfig
  author_pack?: AuthorPackFile | null
  slot_registry_pack?: import('../lib/slotRegistry').SlotRegistryMap | null
  slot_registry_effective?: import('../lib/slotRegistry').SlotRegistryMap | null
  slot_session_overridden_keys?: string[]
  blueprint_groups_pack?: import('../lib/slotRegistry').SlotGroupsMap | null
  dual_core_enabled?: boolean
  pipeline_experimental_actions?: string[]
}

/** `switch_scene` flattened `RoleInfo` fields plus optional scene welcome line */

export async function loadRole(roleId: string): Promise<RoleData> {
  return invokeWithFriendlyError<RoleData>('load_role', { roleId })
}

/** Resolve role asset absolute path; returns `null` when missing (does not throw) */

export async function resolveRoleAssetPath(
  roleId: string,
  relative: string,
): Promise<string | null> {
  return invoke<string | null>('resolve_role_asset_path', {
    roleId,
    relative,
  })
}

export async function readRoleAssetBytes(
  roleId: string,
  relative: string,
): Promise<number[] | null> {
  return invokeWithFriendlyError<number[] | null>('read_role_asset_bytes', {
    roleId,
    relative,
  })
}

/** When `sessionId` is passed, scopes `get_role_info` to that conversation namespace for faster `plugin_backends_*` */

export async function getRoleInfo(
  roleId: string,
  sessionId?: string | null,
): Promise<RoleInfo> {
  return invokeWithFriendlyError<RoleInfo>('get_role_info', {
    req: {
      role_id: roleId,
      session_id: sessionId ?? null,
    },
  })
}


export async function listRoles(): Promise<Array<{ id: string, name: string }>> {
  const rows = await invokeWithFriendlyError<
    Array<{ id: string, name: string, version: string, author: string }>
  >('list_roles', {})
  return rows.map(r => ({ id: r.id, name: r.name }))
}


export async function switchRole(roleId: string): Promise<RoleInfo> {
  return invokeWithFriendlyError<RoleInfo>('switch_role', { roleId })
}


export async function setUserRelation(
  roleId: string,
  relation: string,
): Promise<RoleInfo> {
  return invokeWithFriendlyError<RoleInfo>('set_user_relation', {
    req: { role_id: roleId, relation },
  })
}


export async function setEvolutionFactor(
  roleId: string,
  eventImpactFactor: number,
): Promise<RoleInfo> {
  return invokeWithFriendlyError<RoleInfo>('set_evolution_factor', {
    req: { role_id: roleId, event_impact_factor: eventImpactFactor },
  })
}


export async function setRemoteLifeEnabled(
  roleId: string,
  enabled: boolean,
): Promise<RoleInfo> {
  return invokeWithFriendlyError<RoleInfo>('set_remote_life_enabled', {
    req: { role_id: roleId, enabled },
  })
}


export async function setRoleInteractionMode(
  roleId: string,
  mode: 'immersive' | 'pure_chat',
): Promise<RoleInfo> {
  return invokeWithFriendlyError<RoleInfo>('set_role_interaction_mode', {
    req: { role_id: roleId, mode },
  })
}


export async function setSceneUserRelation(
  roleId: string,
  sceneId: string,
  relation: string,
): Promise<RoleInfo> {
  return invokeWithFriendlyError<RoleInfo>('set_scene_user_relation', {
    req: { role_id: roleId, scene_id: sceneId, relation },
  })
}

/** Remove per-scene identity override; use `set_user_relation` sentinel to restore default identity option */

export async function clearSceneUserRelation(
  roleId: string,
  sceneId: string,
): Promise<RoleInfo> {
  return invokeWithFriendlyError<RoleInfo>('clear_scene_user_relation', {
    req: { role_id: roleId, scene_id: sceneId },
  })
}


export async function exportRolePack(
  roleId: string,
  destPath: string,
): Promise<void> {
  return invokeWithFriendlyError<void>('export_role_pack_command', {
    roleId,
    destPath,
  })
}


export interface RolePackPeek {
  id: string
  name: string
  version: string
}

/** Import role pack: `srcPath` may be `.ocpak`, `.zip`, or extracted directory matching `roles/{id}/` layout */

export async function peekRolePack(srcPath: string): Promise<RolePackPeek> {
  return invokeWithFriendlyError<RolePackPeek>('peek_role_pack_command', {
    srcPath,
  })
}

/** Export role pack to path (ZIP archive or extracted directory) */

export async function importRolePack(
  srcPath: string,
  overwrite: boolean,
): Promise<string> {
  return invokeWithFriendlyError<string>('import_role_pack_command', {
    srcPath,
    overwrite,
  })
}

File: kernel/crates/oclive_kernel_types/src/models/dto/role.rs
//! Role data / summary / info DTOs.

use std::collections::BTreeMap;

use super::super::author_pack::AuthorPackFile;
use super::super::plugin_backends::{
    PluginBackends, PluginBackendsOverride, PluginBackendsSourceMap,
};
use super::super::role::{IdentityBinding, LifeState, PersonalitySource};
use super::super::ui_config::UiConfig;
use serde::{Deserialize, Serialize};

use super::{DisplayMetricsDto, UserRelationDto};
use crate::models::ReplyModeInfoDto;

/// Full role runtime snapshot for settings and plugin panels (`get_role_info`).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoleData {
    pub role_id: String,
    pub name: String,
    pub version: String,
    pub author: String,
    pub description: String,
    /// Whether this role pack contains a valid `adult_extension.json`.
    #[serde(default)]
    pub adult_extension_available: bool,
    /// Present when an invalid adult extension was disabled during load.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub adult_extension_error: Option<String>,
    #[deprecated(note = "use display_metrics.traits")]
    pub personality_vector: Vec<f64>,
    #[deprecated(note = "use display_metrics.favor")]
    pub current_favorability: f64,
    /// UI-only affect snapshot (favor / traits / relation stage).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub display_metrics: Option<DisplayMetricsDto>,
    /// Placeholder; persisted bot emotion deferred to a later milestone.
    pub current_emotion: String,
    pub memory_count: i32,
    pub event_count: i32,
    /// Relation options defined in the role pack.
    pub user_relations: Vec<UserRelationDto>,
    /// Manifest default relation key.
    pub default_relation: String,
    /// Relation state (favorability-driven stage, e.g. Stranger / Friend).
    #[deprecated(note = "use display_metrics.relation_summary")]
    pub relation_state: String,
    /// Current runtime relation (resolved manifest key).
    pub current_user_relation: String,
    pub use_manifest_default: bool,
    /// When remote-presence: whether to generate life trajectory and inner voice (user toggle; persisted in `role_runtime`).
    pub remote_life_enabled: bool,
    /// Role pack `settings.json` → `remote_presence.default_enabled` (suggested mode default); UI hint only.
    pub remote_life_pack_default: Option<bool>,
    /// Effective event impact factor (DB overrides manifest default).
    pub event_impact_factor: f64,
    /// `evolution.personality_source`：`vector` | `profile`
    #[serde(default)]
    pub personality_source: PersonalitySource,
    /// Ollama model used by this role (manifest → `OLLAMA_MODEL` → global default).
    pub effective_ollama_model: String,
    /// Whether identity is scene-bound (manifest `identity_binding`).
    pub identity_binding: IdentityBinding,
    /// Current interaction mode (`role_runtime`).
    pub interaction_mode: String,
    /// Suggested role pack default from `settings.json` (optional).
    pub interaction_mode_pack_default: Option<String>,
    /// Current schedule inference (`null` when unconfigured or no matching time slot).
    #[serde(default)]
    pub current_life: Option<LifeStateDto>,
    /// `settings.json` → `plugin_backends` (matches runtime `PluginHost` resolution).
    #[serde(default)]
    pub plugin_backends: PluginBackends,
    /// Session-level override (current session namespace only; `null` when none).
    #[serde(default)]
    pub plugin_backends_session_override: Option<PluginBackendsOverride>,
    /// Effective backends after session override (runtime panel display and toggle echo).
    #[serde(default)]
    pub plugin_backends_effective: PluginBackends,
    /// Effective backend source (pack/session/env).
    #[serde(default)]
    pub plugin_backends_effective_sources: PluginBackendsSourceMap,
    /// Role pack root `ui.json` (theme, layout, slots, etc.).
    #[serde(default)]
    pub pack_ui_config: UiConfig,
    /// `author.suggested_ui` when non-empty, else same as `pack_ui_config`; plugin UI seed/reset baseline.
    #[serde(default)]
    pub pack_ui_baseline: UiConfig,
    /// Optional full `author.json` (recommended plugins, suggested backends, etc.).
    #[serde(default)]
    pub author_pack: Option<AuthorPackFile>,
    #[serde(default)]
    pub slot_registry_pack: Option<BTreeMap<String, oclive_validation::SlotRegistryEntry>>,
    #[serde(default)]
    pub slot_registry_effective: Option<BTreeMap<String, oclive_validation::SlotRegistryEntry>>,
    #[serde(default)]
    pub slot_session_overridden_keys: Vec<String>,
    /// v2 blueprint `groups` (architecture diagram logical grouping; `null` for legacy).
    #[serde(default)]
    pub blueprint_groups_pack: Option<BTreeMap<String, oclive_validation::SlotGroupEntry>>,
}

/// Lightweight role list entry (`list_roles`).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoleSummary {
    pub id: String,
    pub name: String,
    pub version: String,
    pub author: String,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub featured: bool,
    #[serde(default = "default_preset_order")]
    pub preset_order: u32,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub interaction_mode_suggestion: Option<String>,
    /// Whether this role pack contains a valid `adult_extension.json`.
    #[serde(default)]
    pub adult_extension_available: bool,
    /// Present when an invalid adult extension was disabled during load.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub adult_extension_error: Option<String>,
}

fn default_preset_order() -> u32 {
    999
}

/// Scene id + display label for scene switch UI.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SceneLabelEntry {
    pub id: String,
    pub label: String,
}

/// Current activity inferred from virtual time + manifest `life_schedule` (UI / debug).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LifeStateDto {
    pub label: String,
    pub activity_key: String,
    pub busy_level: f32,
    pub preferred_scene_id: Option<String>,
}

impl From<&LifeState> for LifeStateDto {
    fn from(s: &LifeState) -> Self {
        Self {
            label: s.label.clone(),
            activity_key: s.activity_key.clone(),
            busy_level: s.busy_level,
            preferred_scene_id: s.optional_scene_hint.clone(),
        }
    }
}

/// Role panel / scene UI snapshot (`get_role_info`).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RoleInfo {
    pub role_id: String,
    pub role_name: String,
    pub version: String,
    pub author: String,
    pub description: String,
    /// Whether this role pack contains a valid `adult_extension.json`.
    #[serde(default)]
    pub adult_extension_available: bool,
    /// Present when an invalid adult extension was disabled during load.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub adult_extension_error: Option<String>,
    #[deprecated(note = "use display_metrics.favor")]
    pub current_favorability: f64,
    pub current_emotion: String,
    #[deprecated(note = "use display_metrics.traits")]
    pub personality_vector: Vec<f64>,
    /// UI-only affect snapshot (favor / traits / relation stage).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub display_metrics: Option<DisplayMetricsDto>,
    /// `evolution.personality_source`：`vector` | `profile`
    #[serde(default)]
    pub personality_source: PersonalitySource,
    pub last_interaction: Option<String>,
    /// Available scene ids (manifest + `scenes/` directory).
    pub scenes: Vec<String>,
    /// Same order as `scenes`; `label` from `scenes/{id}/scene.json` `name` or built-in mapping.
    pub scene_labels: Vec<SceneLabelEntry>,
    pub current_scene: Option<String>,
    /// User narrative / send-message context scene (persisted); may differ from `current_scene`.
    pub user_presence_scene: Option<String>,
    /// Virtual world time (UTC ms); 0 if not yet initialized via `get_time_state`.
    pub virtual_time_ms: i64,
    pub user_relations: Vec<UserRelationDto>,
    pub default_relation: String,
    pub current_user_relation: String,
    /// Whether the user selected "default identity" (follows manifest `default_relation`); when true the dropdown should show `OCLIVE_DEFAULT_RELATION_SENTINEL`.
    pub use_manifest_default: bool,
    /// Relation state (`role_runtime.relation_state`).
    #[deprecated(note = "use display_metrics.relation_summary")]
    pub relation_state: String,
    /// Remote inner voice toggle.
    pub remote_life_enabled: bool,
    /// Role pack suggested default for remote inner voice (`settings.json` → `remote_presence.default_enabled`).
    pub remote_life_pack_default: Option<bool>,
    pub event_impact_factor: f64,
    /// Ollama model actually used by this role (manifest → `OLLAMA_MODEL` → global default).
    pub effective_ollama_model: String,
    /// Whether identity is scene-bound (manifest `identity_binding`).
    pub identity_binding: IdentityBinding,
    /// Current interaction mode (`role_runtime`).
    pub interaction_mode: String,
    /// Suggested role pack default from `settings.json` (optional).
    pub interaction_mode_pack_default: Option<String>,
    /// Current schedule inference (`null` when unconfigured or no time window matches).
    #[serde(default)]
    pub current_life: Option<LifeStateDto>,
    /// `settings.json` → `plugin_backends` (consistent with `load_role` / orchestration layer).
    #[serde(default)]
    pub plugin_backends: PluginBackends,
    /// Session-level override (current session namespace only; `null` when none).
    #[serde(default)]
    pub plugin_backends_session_override: Option<PluginBackendsOverride>,
    /// Effective backends after session override (runtime panel display and toggle echo).
    #[serde(default)]
    pub plugin_backends_effective: PluginBackends,
    /// Effective backend source (pack/session/env).
    #[serde(default)]
    pub plugin_backends_effective_sources: PluginBackendsSourceMap,
    /// Whether the currently loaded role has a worldview knowledge index built (`knowledge_index`).
    #[serde(default)]
    pub knowledge_enabled: bool,
    /// `knowledge_index.chunks` count; 0 when index not loaded.
    #[serde(default)]
    pub knowledge_chunk_count: i32,
    /// Role pack root `ui.json` (theme, layout, slots, etc.).
    #[serde(default)]
    pub pack_ui_config: UiConfig,
    /// `author.suggested_ui` when non-empty, else same as `pack_ui_config`; plugin UI seed/reset baseline.
    #[serde(default)]
    pub pack_ui_baseline: UiConfig,
    /// Optional `author.json`.
    #[serde(default)]
    pub author_pack: Option<AuthorPackFile>,
    /// Role pack `slot_registry` (v2 blueprint; `null` for legacy packs).
    #[serde(default)]
    pub slot_registry_pack: Option<BTreeMap<String, oclive_validation::SlotRegistryEntry>>,
    /// Effective `slot_registry` after session overrides.
    #[serde(default)]
    pub slot_registry_effective: Option<BTreeMap<String, oclive_validation::SlotRegistryEntry>>,
    /// Instance keys with overrides in the current session.
    #[serde(default)]
    pub slot_session_overridden_keys: Vec<String>,
    /// v2 blueprint `groups` (architecture diagram logical grouping; `null` for legacy).
    #[serde(default)]
    pub blueprint_groups_pack: Option<BTreeMap<String, oclive_validation::SlotGroupEntry>>,
    /// `runtime_config.dual_core.enabled` and `pipeline.experimental` is non-empty.
    #[serde(default)]
    pub dual_core_enabled: bool,
    /// `pipeline.experimental` action list (architecture diagram / debug read-only).
    #[serde(default)]
    pub pipeline_experimental_actions: Vec<String>,
    /// Effective reply post-processor enabled (role pack + distro merge).
    #[serde(default)]
    pub reply_post_processor_enabled: bool,
    /// `builtin` | `remote` | `directory` | `off` when disabled.
    #[serde(default)]
    pub reply_post_processor_backend: String,
    /// Effective builtin profile (`standard` | `minimal`) when enabled.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reply_post_processor_profile: Option<String>,
    /// Effective reply presentation mode; present when the role pack enables segmented replies.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub pack_reply_mode: Option<ReplyModeInfoDto>,
}
